core/domain/exp_services_test.py

```diff
@@ -881,13 +881,102 @@ def test_export_by_versions(self):
 def _get_change_list(state_name, property_name, new_value):
     """Generates a change list for a single state change."""
     return [{
-        'cmd': 'edit_state_property',
+        'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
         'state_name': state_name,
         'property_name': property_name,
         'new_value': new_value
     }]
 
 
+class UpdateGadgetTests(ExplorationServicesUnitTests):
+    """Test updating a gadget."""
+
+    def setUp(self):
+        super(UpdateGadgetTests, self).setUp()
+        exploration = self.save_new_valid_exploration(
+            self.EXP_ID, self.owner_id)
+
+        self.init_state_name = exploration.init_state_name
+
+        self.param_changes = [{
+            'customization_args': {
+                'list_of_values': ['1', '2'], 'parse_with_jinja': False
+            },
+            'name': 'myParam',
+            'generator_id': 'RandomSelector'
+        }]
+        # List of answer groups to add into an interaction.
+        self.interaction_answer_groups = [{
+            'rule_specs': [{
+                'rule_type': 'Equals',
+                'inputs': {'x': 0},
+            }],
+            'outcome': {
+                'dest': self.init_state_name,
+                'feedback': ['Try again'],
+                'param_changes': []
+            },
+        }]
+        # Default outcome specification for an interaction.
+        self.interaction_default_outcome = {
+            'dest': self.init_state_name,
+            'feedback': ['Incorrect', '<b>Wrong answer</b>'],
+            'param_changes': []
+        }
+
+
+    def test_gadget_commands(self):
```

Comment:
Hello @seanlip i have done that but the exploration.add_gadget() function isnt adding the gadget, the test gives an error `ValueError: Gadget ATestGadget does not exist.`. This is a sample code ` def test_delete_gadget_cmd(self):
        """test delete"""
        TEST_GADGET_CUSTOMIZATION_ARGS = {
            'adviceObjects': {
                'value': [{
                    'adviceTitle': 'b',
                    'adviceHtml': '<p>c</p>'
                }]
            }
        }

        TEST_GADGET_DICT = {
            'gadget_type': 'TestGadget',
            'gadget_name': 'ATestGadget',
            'customization_args': TEST_GADGET_CUSTOMIZATION_ARGS,
            'visible_in_states': ['First state']
        }

        exploration = exp_services.get_exploration_by_id(self.EXP_ID)
        exploration.add_gadget(TEST_GADGET_DICT, 'bottom')

        self.assertIn('ATestGadget', exploration.get_all_gadget_names())

        exp_services.update_exploration(self.owner_id, self.EXP_ID, [{
            'cmd': exp_domain.CMD_DELETE_GADGET,
            'gadget_name': 'ATestGadget',
            }], 'delete a gadget')

        exploration = exp_services.get_exploration_by_id(self.EXP_ID)

        self.assertNotIn('ATestGadget', exploration.get_all_gadget_names())`